Makefile

```diff
@@ -48,4 +48,37 @@ submodules:
 
 # We must ensure up is installed in tool cache prior to build as including the k8s_tools machinery prior to the xpkg
 # machinery sets UP to point to tool cache.
-build.init: $(UP)
\ No newline at end of file
+build.init: $(UP)
+
+# ====================================================================================
+# End to End Testing
+
+KIND_VERSION = v0.16.0
+KIND_CLUSTER_NAME ?= uptest
+PROVIDER_GCP_VERSION ?= v0.15.0
+PROVIDER_HELM_VERSION ?= v0.12.0
+
+controlplane.up: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) setting up controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 || $(KIND) create cluster --name=$(KIND_CLUSTER_NAME)
+	@$(KUBECTL) -n upbound-system get cm universal-crossplane-config >/dev/null 2>&1 || $(UP) uxp install
+	@$(KUBECTL) -n upbound-system wait deploy crossplane --for condition=Available --timeout=120s
+	@$(KUBECTL) -n upbound-system create secret docker-registry package-pull-secret --docker-server=xpkg.upbound.io --docker-username=${UPBOUND_DOCKER_USERNAME} --docker-password=${UPBOUND_DOCKER_PASSWORD} -o yaml --dry-run=client | $(KUBECTL) apply -f -
+	@$(KUBECTL) get provider.pkg upbound-provider-gcp > /dev/null 2>&1 || $(UP) ctp provider install upbound/provider-gcp:$(PROVIDER_GCP_VERSION) --package-pull-secrets=package-pull-secret
```

Comment:
all logic around `package-pull-secret` can be removed as we are not gating official providers anymore